raft topology: send cdc generation data in parts

margdoc opened the pull request:
The CDC generation data can be large and not fit in a single command. This pr splits it into multiple mutations by smartly picking a `mutation_size_threshold` and sending each mutation as a separate group 0 command.

Commands are sent sequentially to avoid concurrency problems. 

Topology snapshots contain only mutation of current CDC generation data but don't contain any previous or future generations. If a new generation of data is being broadcasted but hasn't been entirely applied yet, the applied part won't be sent in a snapshot. New or delayed nodes can never get the applied part in this scenario.

Send the entire cdc_generations_v3 table in the snapshot to resolve this problem.

A mechanism to remove old CDC generations will be introduced as a follow-up.

margdoc commented:
@kbr-scylla please review

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1352/

kbr-scylla commented on service/raft/group0_state_machine.cc:
```diff
@@ -108,6 +108,9 @@ future<> group0_state_machine::apply(std::vector<raft::command_cref> command) {
            return _ss.topology_transition(_sp, _cdc_gen_svc, cmd.creator_addr, std::move(chng.mutations));
+        },
+        [&] (write_mutations& muts) -> future<> {
+           return _ss.write_topology_mutations(_sp, cmd.creator_addr, std::move(muts.mutations));
```
The mutations sent in `write_mutations` don't necessarily have to be "topology mutations". In fact the mutations that you're sending with this command in later commits - I wouldn't really call them topology mutations, they don't affect the `topology` struct or `system.topology` table. They write data to CDC generations table (`system.cdc_generations`), and only the last command updates `topology` to point to that data.

Furthermore, it's not safe in general to split topology mutations (those that may modify `topology` struct) and commit them to group 0 in multiple commands. Because after applying once of them, a node may crash and then restart. Then it will call `topology_state_load` and observe partial state.

The CDC mutations don't have this problem. If we crash after committing only a subset of them, it won't affect anything other than storing some garbage data in `system.cdc_generations` that isn't referenced by anything. Only the last command sent by the coordinator updates the `system.topology` table and the `topology` struct which stores a pointer (key) to `system.cdc_generations` which contains the entire generation data at this moment.

So in fact the mutations sent by `write_mutations` not only aren't necessarily topology mutations - they must not be topology mutations, unless they contain a full update of topology state after which it's safe to call `topology_state_load`. But then you'd use `topology_change` command instead of `write_mutations`. So `write_mutations` are for purposes different than updating topology.

Having said all that, I propose the following changes:
- write a comment at `struct write_mutations` (in group0_state_machine.hh) explaining that this command is used to write data to tables other than topology or schema tables and it doesn't update any in-memory data structures
- rename `write_topology_mutations` function to something more generic, maybe simply `write_mutations`. I would also put it outside `storage_service`, maybe in group0_state_machine.cc (it looks like it could be a free-standing function, not a member). Then `storage_service::topology_transition` won't be able to call it, so put the call also in group0_state_machine in the `topology_change` case above the call to `topology_transition`. You should then specify at `topology_transition` declaration site the precondition that the topology mutations were already written to disk; the function only transitions the in-memory state machine.

margdoc replied:
Done.

kbr-scylla commented on cdc/generation.cc:
```diff
@@ -272,8 +273,10 @@ future<utils::chunked_vector<mutation>> get_cdc_generation_mutations(
     res.emplace_back(s, partition_key::from_singular(*s, id));
     res.back().set_static_cell(to_bytes("num_ranges"), int32_t(desc.entries().size()), ts);
     size_t size_estimate = 0;
+    size_t result_size_estimate = 0;
```
I'd call it `total_size_estimate`

margdoc replied:
Done.

kbr-scylla commented on cdc/generation.cc:
```diff
@@ -292,6 +295,21 @@ future<utils::chunked_vector<mutation>> get_cdc_generation_mutations(
+    utils::get_local_injector().inject("cdc_generation_mutations_replication", [&res, result_size_estimate, mutation_size_threshold] {
+        utils::chunked_vector<mutation> new_res;
+
+        size_t number_of_copies = (mutation_size_threshold / result_size_estimate + 1) * 2;
```
Please write a comment explaining what this calculation aims to achieve. (The commit message provides a nice explanation; but the commit message is lost in code)

margdoc replied:
Done.

kbr-scylla commented:
Please also rebase the PR on `master` - https://github.com/scylladb/scylladb/pull/13683 got merged

margdoc commented:
v2:
* rebased on master

margdoc commented:
v3:
* s/result_size_estimate/total_size_estimate
* Added a comment to the "cdc_generation_mutations_replication" injection
* Moved `write_topology_mutations` to `group0_state_machine.cc` and remove call to it in `topology_transition`
* Added precondition to `topology_transition` and removed unnecessary arguments
* s/write_topology_mutations/write_mutations_to_database
* Added a comment to the "write_mutations" struct

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1494/

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -938,6 +931,45 @@ class topology_coordinator {
+                    group0_command g0_cmd = _group0.client().prepare_command(std::move(change), guard, reason);
+                    co_await _group0.client().add_entry(std::move(g0_cmd), std::move(guard));
+                    done = true;
+                } catch (group0_concurrent_modification&) {
```
Please write a comment providing the reason why we catch it here instead of letting it propagate to the top-level of coordinator code (e.g. we don't want to start from scratch if there are N commands to send and there's a concurrent schema change when we're nearly done - you should calculate what N will be in large clusters).

margdoc replied:
Done.

kbr-scylla commented on cdc/generation.cc:
```diff
@@ -25,6 +25,7 @@
 #include "gms/inet_address.hh"
```
Re Commit message: please describe how the test fails without the changes from this PR (i.e. if we try to send the data in a single command).

margdoc replied:
Done.

kbr-scylla commented on cdc/generation.cc:
```diff
@@ -292,6 +295,22 @@ future<utils::chunked_vector<mutation>> get_cdc_generation_mutations(
 
+    total_size_estimate += size_estimate;
+
+    // Copy mutations n times, where n is picked that the memory size of all mutations exceeds `max_command_size`.
```
"is picked that" -> "is picked so that"
"the memory size of all mutations exceeds" -> "the memory size of all mutations together exceeds" (to make it clear that it's not each mutation size exceeding `max_command_size`; it's their total)

margdoc replied:
Done.

kbr-scylla commented on test/topology_experimental_raft/test_cdc_generation_data.py:
```diff
@@ -0,0 +1,19 @@
+    first_server = await manager.server_add(config=config)
+
+    async with inject_error(manager.api, first_server.ip_addr, 'cdc_generation_mutations_replication'):
+        for _ in range(3):
```
Isn't it enough to boot a single node after doing the injection? If it is, then let's do only that so that the test doesn't run unnecessarily long.

margdoc replied:
Good point, changed.

kbr-scylla commented on test/topology_experimental_raft/test_cdc_generation_data.py:
```diff
@@ -0,0 +1,19 @@
+import pytest
+
+@pytest.mark.asyncio
+async def test_send_data_in_parts(manager: ManagerClient):
```
Please provide a Python docstring ("""blah blah blah""") describing how the test works (i.e.: it's a regression test, the injection forces the topology coordinator to send CDC generation data in multiple parts, if it didn't the command size would go over commitlog segment size limit making it impossible to commit and apply the command)

margdoc replied:
Done.

margdoc commented:
v4:
* Added more comments
* Added a description of the proposed solution in the commit message
* Boot only one node in the test

margdoc commented on service/storage_service.cc:
```diff
@@ -20,6 +20,10 @@
 #include "db/system_distributed_keyspace.hh"
```
Done.

margdoc commented on service/storage_service.cc:
```diff
@@ -930,7 +930,8 @@ class topology_coordinator {
         // was done when the mutations were stored in a regular distributed table.
-        const size_t mutation_size_threshold = 2'000'000;
+        const size_t max_command_size = _raft.max_command_size();
+        const size_t mutation_size_threshold = max_command_size / 2;
```
Done.

margdoc commented on service/raft/group0_state_machine.hh:
```diff
@@ -37,8 +37,14 @@ struct topology_change {
 };
 
+// This command is used to write data to tables other than topology or schema tables
+// and it doesn't update any in-memory data structures
```
Done.

kbr-scylla commented on service/raft/group0_state_machine.cc:
```diff
@@ -57,6 +58,22 @@ static mutation convert_history_mutation(canonical_mutation m, const data_dictio
     return m.to_mutation(db.find_schema(db::system_keyspace::NAME, db::system_keyspace::GROUP0_HISTORY));
 }
 
+future<> write_mutations_to_database(storage_proxy& proxy, gms::inet_address from, std::vector<canonical_mutation> cms) {
```
Nit: `static`

margdoc replied:
Done.

kbr-scylla commented on service/raft/group0_state_machine.cc:
```diff
@@ -57,6 +58,22 @@ static mutation convert_history_mutation(canonical_mutation m, const data_dictio
+        }
+    } catch (replica::no_such_column_family& e) {
+        slogger.error("Error while applying topology mutations from {}: {}", from, e);
+        throw std::runtime_error(::format("Error while applying topology mutations: {}", e));
```
The messages mention "topology mutations" but the function will be used for other mutations too

margdoc replied:
Changed.

kbr-scylla commented on idl/group0_state_machine.idl.hh:
```diff
@@ -27,8 +27,12 @@ struct topology_change {
     std::vector<canonical_mutation> mutations;
```
> With this command, it's possible to split `topology_change` command into parts
if it doesn't fit max command size.

https://github.com/scylladb/scylladb/pull/13962#discussion_r1208472786

gleb-cloudius replied:
But how is it possible to split it atomically? Is atomicity is not needed?

kbr-scylla replied:
> But how is it possible to split it atomically? Is atomicity is not needed?

That's why I asked for updating the commit message and elaborating, @margdoc.

kbr-scylla replied:
@gleb-cloudius to answer your question: `write_mutations` will be used to persist CDC generation data, which doesn't have to be atomic. If we break in the middle it's fine, we'll just have some garbage partition in the table that nobody uses for anything. If we commit all parts then the last command, `topology_change`, will update the topology state which stores the CDC generation UUID which is a key to a partition in that table which at that point contains all the CDC generation data.

gleb-cloudius replied:
But them may be we should not guard those writes with state id? Should they be strictly ordered? May be we can write them not over raft at all?

kbr-scylla replied:
Writing them over Raft guarantees that the last `topology_change` command will observe all that data on every node where it's executed (consistency across configuration changes and so on). So I think it's the simplest thing to use Raft for sending those commands.

Regarding state IDs, indeed we might not need them.

Hmm, now that you mentioned this I see a correctness problem... When we're pulling snapshots, we only pull CDC generation data for the last generation UUID that's saved in topology. So if we start sending data for new generation, then another node joins or some slow node requires a snapshot transfer, that node may not get all this data and we're in trouble. @margdoc we need to think how to fix that.

margdoc replied:
> https://github.com/scylladb/scylladb/pull/13962#discussion_r1208472786

Fixed.

> Hmm, now that you mentioned this I see a correctness problem... When we're pulling snapshots, we only pull CDC generation data for the last generation UUID that's saved in topology. So if we start sending data for new generation, then another node joins or some slow node requires a snapshot transfer, that node may not get all this data and we're in trouble. @margdoc we need to think how to fix that.

We decided to send mutations of CDC generation data that ate being broadcasted but aren't entirely applied yet, in the topology snapshot.

kbr-scylla commented on service/raft/group0_state_machine.hh:
```diff
@@ -37,8 +37,13 @@ struct topology_change {
     std::vector<canonical_mutation> mutations;
 };
 
+// This command is used to write data to tables over raft.
```
Please provide more explanation in the comment as described here:
https://github.com/scylladb/scylladb/pull/13962#discussion_r1200645161
specifically, this part:
> write a comment at struct write_mutations (in group0_state_machine.hh) explaining that this command is used to write data to tables **other than topology or schema tables and it doesn't update any in-memory data structures**

margdoc replied:
Fixed.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -20,6 +20,10 @@
 #include "db/system_distributed_keyspace.hh"
 #include "db/consistency_level.hh"
```
> With the default commitlog segment size, `mutation_size_threshold` will be 4 MB.
In large clusters e.g. 100 nodes, 64 shards per node, 256 vnodes cdc generation
data can reach the size of 30 MB, thus

> In a multi-DC cluster with 100ms latencies between DCs, which means ~200ms for
each command, because we first perform a barrier in start_operation and then commit
the command. Thus this operation will take about 1.6s which is incomparably shorter
than bootstrapping operation.

The "thus" in the first paragraph ends it abruptly, is something missing there?

BTW. bootstrapping is quick if there is no data in the cluster - but usually if one has 100 nodes they have tons of data, so indeed streaming/repair will take much longer (hours/days) and additional 1.6s is not a problem.

margdoc replied:
> The "thus" in the first paragraph ends it abruptly, is something missing there?

Yes, thanks, I missed it. Now it is fixed.

kbr-scylla commented:
@gleb-cloudius @piodul please review

gleb-cloudius commented on service/raft/group0_state_machine.cc:
```diff
@@ -36,6 +36,7 @@
 #include "timestamp.hh"
 #include "utils/overloaded_functor.hh"
 #include <optional>
+#include "replica/database.hh"
```
Typo in the subject: appling->applying.

margdoc replied:
Fixed.

avikivity commented:
> The CDC generation data can be large and not fit in a single command. This pr splits it into multiple mutations by smartly picking a `mutation_size_threshold` and sending each mutation as a separate group 0 command.
> 
> Commands are sent sequentially to avoid concurrency problems.

What happens if one command succeeds and the next fails? Do we end up in an inconsistent state?

piodul commented on cdc/generation.cc:
```diff
@@ -292,6 +295,22 @@ future<utils::chunked_vector<mutation>> get_cdc_generation_mutations(
+
+        size_t number_of_copies = (mutation_size_threshold / total_size_estimate + 1) * 2;
+        for (size_t i = 0; i < number_of_copies; ++i) {
+            for (auto r : res) {
+                new_res.push_back(r);
+            }
```
I know this is test code, but... you have an unnecessary copy here that you could remove. Either:

- change the type of `r` to `const auto&`,
- `std::move(r)` when `pushing_back`,
- or replace the innermost loop with `std::copy`.

margdoc replied:
Changed to `std::copy`.

piodul commented on service/storage_service.cc:
```diff
@@ -443,23 +443,7 @@ future<> storage_service::topology_state_load(cdc::generation_service& cdc_gen_s
 }
 
-future<> storage_service::topology_transition(storage_proxy& proxy, cdc::generation_service& cdc_gen_svc, gms::inet_address from, std::vector<canonical_mutation> cms) {
-    assert(this_shard_id() == 0);
```
I think this assertion should stay. It's still true that `topology_transition` needs to be called on shard 0.

margdoc replied:
Fixed.

piodul commented on service/storage_service.cc:
```diff
@@ -946,17 +987,15 @@ class topology_coordinator {
 
             auto tmptr = get_token_metadata_ptr();
-            auto [gen_uuid, gen_mutations] = co_await prepare_new_cdc_generation_data(tmptr, guard, std::nullopt);
+            auto [gen_uuid, guard_, mutation] = co_await prepare_and_broadcast_cdc_generation_data(tmptr, std::move(guard), std::nullopt);
```
Hm, this function operates on the `guard` variable, but you are introducing `guard_` here. I think it's slightly error prone and it would be better to reuse the `guard` variable - basically, you would like to put the new guard there.

How about std::moving the guard_ into guard after this line?

margdoc replied:
Good idea, fixed.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1669/

margdoc commented on service/storage_service.cc:
```diff
@@ -938,6 +931,54 @@ class topology_coordinator {
+    // Returns the generation's UUID, last created guard and last mutation which
+    // will be sent with additional topology data by the caller.
+    //
+    // With the default commitlog segment size, `mutation_size_threshold` will be 4 MB. In large clusters e.g.
+    // 100 nodes, 64 shards per node, 256 vnodes cdc generation data can reach the size of 30 MB, thus
```
Yes, thanks, I missed it. Now it is fixed.

margdoc commented on service/storage_service.cc:
```diff
@@ -938,6 +931,54 @@ class topology_coordinator {
+
+        std::vector<canonical_mutation> updates{gen_mutations.begin(), gen_mutations.end()};
+
+        for (size_t i{0}; i < updates.size() - 1; ++i) {
```
There wasn't any specific reason, it was a matter of habit. Fixed.

margdoc commented on service/storage_service.cc:
```diff
@@ -938,6 +931,54 @@ class topology_coordinator {
+                    done = true;
+                } catch (group0_concurrent_modification&) {
+                    // We don't want to start from scratch if there are 8 commands to send and
+                    // there's a concurrent schema change when we're nearly done.
```
Now `write_mutations` are sent unguarded, so this code is deleted. This was explained in the comment that was missing.

margdoc commented:
v6:
* extended topology snapshot
* fixed missing comment and commit message
* expanded some comments
* fixed `write_mutations_to_database` comment
* added `static` to `write_mutations_to_database`
* fixed typos
* `guard_` -> `guard`
* changed copying mutations in injection to `std::copy`
* restored assertion in `topology_transition`

margdoc commented:
v7:
* rebased on `master`

margdoc commented:
@avikivity
> What happens if one command succeeds and the next fails? Do we end up in an inconsistent state?

`write_mutations` only send mutations that will be applied to the database, it doesn't change in-memory structures. If not all commands will be sent, this means that the last one (`topology_change`) also won't be sent, thus the state won't be reloaded and nodes end up with garbage in the `cdc_generations_v3` table.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2034/

avikivity commented:
> @avikivity
> > What happens if one command succeeds and the next fails? Do we end up in an inconsistent state?
> 
> `write_mutations` only send mutations that will be applied to the database, it doesn't change in-memory structures. If not all commands will be sent, this means that the last one (`topology_change`) also won't be sent, thus the state won't be reloaded and nodes end up with garbage in the `cdc_generations_v3` table.

Sounds like a disaster.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -20,6 +20,10 @@
 #include "db/system_distributed_keyspace.hh"
 #include "db/consistency_level.hh"
 #include <seastar/core/smp.hh>
+#include "mutation/canonical_mutation.hh"
```
> In a multi-DC cluster with 100ms latencies between DCs, which means ~200ms for
each command, because we first perform a barrier in start_operation and then commit
the command. Thus this operation will take about 1.6s which is incomparably shorter
than bootstrapping operation (bootstrapping is quick if there is no data in the cluster,
but usually if one has 100 nodes they have tons of data, so indeed streaming/repair
will take much longer (hours/days)).

With tablets the bootstrap time will not depend on the amount of data. A node will be added without streaming and data will be migrated to it later by a load balancer.

> The alternative solution to sequential sending could be sending these commands
concurrently. But it would be complex, it's not clear how to do it with `group0_guard`
and this operation's time is not a problem.

Now the data is send unguarded, so is there still a problem?

I would even do writes through storage proxy with CL=ALL (retrying each failed write individually). Well this should not be a regular CL=ALL (since we may want to ignore some nodes), but something that gets a list of nodes and use them as CL=ALL targets.

gleb-cloudius replied:
> Each mutation is sent in separate raft command and is retried as long as `group0_concurrent_modification` occurs.

This is no longer the case since the exception is not thrown any longer, no?

gleb-cloudius replied:
> > Now the data is send unguarded, so is there still a problem?
> 
> We shouldn't send all mutations at once, we should limit concurrency based on the number of nodes, size and number of mutations. That's still a bit of complexity.

We have `max_concurrent_for_each`, but I do not understand why should we limit concurrency here especially dynamically.

> 
> I would prefer we do it the simpler way now (sequentially). It can be optimized in the future if we really think that the <=2 seconds is so problematic. But it happens only in very large clusters anyway.
> 
> > I would even do writes through storage proxy with CL=ALL (retrying each failed write individually). Well this should not be a regular CL=ALL (since we may want to ignore some nodes), but something that gets a list of nodes and use them as CL=ALL targets.
> 
> Every member of group 0 must observe the data inside CDC_GENERATIONS_V3 when it gets the `topology_change` that commits the generation. So you need to ensure that your CL=ALL targets include exactly every member of group 0. Furthermore, that data needs to be streamed to new nodes when they join group 0, and if we didn't make it part of group 0 as you suggest (IIUC) then we would have to stream it **before** they join group 0 (because as soon as they join, their topology state points to the data in the table).

If you do it with CL=ALL and retry each failure it is guarantied that all nodes will have it when `topology_change`  runs. I am not sure streaming is needed. We do not (will not) add nodes while we add nodes. Wrote it in another comment.

> 
> So why not do it like in this PR, and keep the data in group 0, which solves all these problems automagically? If you suggest an alternative you should explain why, and what are the downsides of current solution that your alternative solves.

This `write_mutation` command looks like a one use hack. If the goal is serialization there are other ways.

gleb-cloudius replied:
> > We have max_concurrent_for_each, but I do not understand why should we limit concurrency here especially dynamically.
> 
> Limiting memory consumption - e.g. suppose you have 100 nodes and 30MB generation, if you send the write concurrently to everyone you could have a memory usage spike up to 3GB IIUC.

This is the job of raft to limit the consumption given that we use raft (and raft does limit log length). If we move to storage proxy a static concurrency is sufficient. Sending one mutation at a time is static concurrency of one anyway. 

> 
> > If you do it with CL=ALL and retry each failure it is guarantied that all nodes will have it when topology_change runs. I am not sure streaming is needed. We do not (will not) add nodes while we add nodes.
> 
> New nodes will join group 0 later. They will need to access this data too. So let's distribute it through group 0. In general if something in group 0 relies on some data it's simplest to put that data in group 0.

They do not need partial data, and last generation we already transfer as part of the snapshot.

> 
> Reading the data even by the current members is simplified when it's written using group 0, because then it is living in a local table and to obtain it, we just do a local read. We don't need to do CL=ONE distributed reads like we currently do with CDC_GENERATIONS_V2, which requires retry loops and so on.

I propose to write it into the new table, just use storage proxy code for that. I clarified that when I say CL=ALL I do not mean real CL=ALL:

"Well this should not be a regular CL=ALL (since we may want to ignore some nodes), but something that gets a list of nodes and use them as CL=ALL targets."

I propose to write the code similar to what we have for batches in storage_proxy now. It will construct specially crafted mutation_write_handler with all the targets that we want to write into.

kbr-scylla replied:
> They do not need partial data, and last generation we already transfer as part of the snapshot.

This data becomes the "last generation" once we send all of it.

kbr-scylla replied:
> I propose to write the code similar to what we have for batches in storage_proxy now. It will construct specially crafted mutation_write_handler with all the targets that we want to write into.

Ah, and implementing special storage_proxy logic for being able to do distributed writes to local tables with a specially crafted `mutation_write_handler` with selected targets is simpler than adding a `write_mutations` command. I see.

gleb-cloudius replied:
> > I propose to write the code similar to what we have for batches in storage_proxy now. It will construct specially crafted mutation_write_handler with all the targets that we want to write into.
> 
> Ah, and implementing special storage_proxy logic for being able to do distributed writes to local tables with a specially crafted `mutation_write_handler` with selected targets is simpler than adding a `write_mutations` command. I see.

It is. The code will be short (handlers are extensible to support such use case). It will be more efficient. And it will not require all the dense here with ` in_progress_cdc_generation_id` here. Why do you directly assume it is more complicated?

kbr-scylla replied:
First of all, "complicated" doesn't mean that it requires boilerplate. "Complicated" means you have to think harder on why a particular solution is correct - at least that's the most important consideration IMO.

So if we have two writes (one for writing data, one for commiting CDC generation) which need to go to all group 0 members, you do the second through group 0 command, and the second depends on the first, how do we ensure that the second observes the first? The obvious answer is to do the first write through group 0 too.

The nonobvious answer is to do the first write using a separate mechanism, then rely on a fact which will be hopefully documented in a comment somewhere, that we do not join new members to group 0 while we're creating a new CDC generation, and then pray that this assumption never changes in the future. Actually even now I'm not sure if this assumption is true, it probably isn't to be honest, because we can call the "repair CDC generation API" and concurrently start joining a new node and this new group 0 member needs to get the CDC genetration created by the API.

So at least considering which solution is easier to understand and prove correct - IMO my solution is simpler.

Regarding boilerplate, there is a very small amount of it in this solution, as you can see in the PR - add a new command type (which we may find useful in the future too, e.g. for broadcast tables) and a new field to topology state.

In your solution, I am not able to estimate how much boilerplate will be needed. And it's also not simple boilerplate, because we invent new things like being able to do distributed writes to a local replication strategy table, which is pretty confusing.

kbr-scylla replied:
So IIUC the correctness argument you provide is -- we do global RPCs from the coordinator already, and this must be correct or our design is broken, so let's also introduce a way to do a global write through `storage_proxy` and use it, it will be correct by extension.

My correctness argument is - both writes are done through group 0 so the second must observe the first.

Maybe we need an external arbiter to decide which argument is simpler.

And you didn't address my amount-of-boilerplate concern. It's hard for me to believe that extending storage_proxy with this new mechanism will require less boilerplate than using `write_mutations`.

gleb-cloudius replied:
> So IIUC the correctness argument you provide is -- we do global RPCs from the coordinator already, and this must be correct or our design is broken, so let's also introduce a way to do a global write through `storage_proxy` and use it, it will be correct by extension.

Global write is not different then global RPC. Fencing rpc also does writes. My correctness argument is that we design the system to be correct this way and any additional mechanism is unnecessary complication. Stick to the design.

> 
> My correctness argument is - both writes are done through group 0 so the second must observe the first.

Why not do everything through grioup0 them. We _can_ do RPCs through it. It is just unnecessary complication. We do not need order between CDC table writes, so why do you insist writing them through raft just in case?

> 
> Maybe we need an external arbiter to decide which argument is simpler.
> 
> And you didn't address my amount-of-boilerplate concern. It's hard for me to believe that extending storage_proxy with this new mechanism will require less boilerplate than using `write_mutations`.

Because we will not know until we try, but we have a lot of specialized write handlers (batch, views, hints, paxos), so the code is designed to be extendible (albeit  through an awkward templated methods).

kbr-scylla replied:
> Global write is not different then global RPC. 

In general it is different. A global write must propagate during topology changes too. Global RPC is only done on the current set of members.

In this case you made the point that we don't need to propagate this particular global write during topology changes, so we can reduce it to a global RPC.

> Why not do everything through grioup0 them. We can do RPCs through it. It is just unnecessary complication. 

I would actually prefer doing it that way if we could, but we can't due to one reason - snapshots. Because of snapshots there is no guarantee that every node will call `apply()`.

This is another difference between a global write and a global RPC. Global write can be propagated through a global RPC for some nodes and by a snapshot for others.

But this is off-topic.

> We do not need order between CDC table writes, so why do you insist writing them through raft just in case?

We **do** need order but not between the CDC table writes within each other, we need order between the set of CDC table writes as a whole and the last `topology_change` which updates `topology`, pointing to the data that was written by the CDC table writes. And because of that order requirement, I claim using group 0 for those writes is just the most obvious thing.

gleb-cloudius replied:
> > Global write is not different then global RPC.
> 
> In general it is different. A global write must propagate during topology changes too. Global RPC is only done on the current set of members.
> 
> In this case you made the point that we don't need to propagate this particular global write during topology changes, so we can reduce it to a global RPC.

Exactly. IMO this is what allows for simplification.

> 
> > Why not do everything through grioup0 them. We can do RPCs through it. It is just unnecessary complication.
> 
> I would actually prefer doing it that way if we could, but we can't due to one reason - snapshots. Because of snapshots there is no guarantee that every node will call `apply()`.
> 
> This is another difference between a global write and a global RPC. Global write can be propagated through a global RPC for some nodes and by a snapshot for others.
> 
> But this is off-topic.

It is. Otherwise we could have argue how RPCs can be encoded in a snapshot as well :)

> 
> > We do not need order between CDC table writes, so why do you insist writing them through raft just in case?
> 
> we need order between the set of CDC table writes as a whole and the last `topology_change`

Exactly my point. So why not do those set of write in most efficient way (since no order is needed) and make sure that there is an order between side effect of those writes and last `topology_change` command like we do for any other RPC with side effects?

kbr-scylla replied:
>  So why not do those set of write in most efficient way (since no order is needed) 

If that's the problem, we can do it through group 0 too, since they are unguarded, we can send these commands concurrently.

BTW. the global RPCs we do from the coordinator currently are only contacting NORMAL nodes, which is probably a bug - we should be contacting bootstrapping node(s) as well. Definitely if we do the CDC generation write through a global RPC we need to do it on bootstrapping node too.

Ayyy More complications. Why not just do the write through group 0 like it's supposed to be used...

gleb-cloudius replied:
> > So why not do those set of write in most efficient way (since no order is needed)
> 
> If that's the problem, we can do it through group 0 too, since they are unguarded, we can send these commands concurrently.
> 

That what I asked in the review. But storage proxy will still be more efficient in multi-dc setup.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1074,6 +1067,51 @@ class topology_coordinator {
+        std::vector<canonical_mutation> updates{gen_mutations.begin(), gen_mutations.end()};
+
+        if (updates.size() > 1) {
+            guard.~group0_guard();
```
I would rather have something like `release_guard()`.

margdoc replied:
Done.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1074,6 +1067,51 @@ class topology_coordinator {
+                slogger.trace("raft topology: do update {} reason {}", updates[i], reason);
+                write_mutations change{{std::move(updates[i])}};
+                group0_command g0_cmd = _group0.client().prepare_command(std::move(change));
+                co_await _group0.client().add_entry_unguarded(std::move(g0_cmd));
```
Better to check that term did not change between iteration to not do useless work.

margdoc replied:
`write_mutations` are sent in parallel now, so checking that the term did not change is pointless.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -5398,39 +5426,48 @@ void storage_service::init_messaging_service(sharded<service::storage_proxy>& pr
             co_return raft_topology_snapshot{
                 .topology_mutations = std::move(topology_mutations),
                 .cdc_generation_mutation = std::move(cdc_generation_mutation),
+                .cdc_in_progress_generation_mutation = std::move(cdc_in_progress_generation_mutation),
```
Why do you need a separate field for it?

margdoc replied:
This part of the code was removed due to a change of topology snapshots. Closing conversation.

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -258,6 +258,7 @@ schema_ptr system_keyspace::topology() {
             .with_column("ignore_msb", int32_type)
             .with_column("supported_features", set_type_impl::get_instance(utf8_type, true))
             .with_column("new_cdc_generation_data_uuid", uuid_type, column_kind::static_column)
+            .with_column("in_progress_cdc_generation_id", uuid_type, column_kind::static_column)
```
> Topology snapshots contain only mutation of current CDC generation data but don't
contain any previous or future generations. If new a generation of data is being
broadcasted but hasn't been entirely applied yet, the applied part won't be sent
in a snapshot. In this scenario, new or delayed nodes can never get the applied part.

So how this can happen? If we broadcast new CDC generation it means we either bootstrap a node or execute "cdc generation recalculation" command. In both cases we cannot join another node until the operation finishes (either succeeds or fails).

Well, right now a new node joins group0 and adds itself to tho topology and that may happen while another operation is in progress, but we are changing it. The coordinator will add nodes to group0 only after previous operation changes, so I think all this complexity will not be needed.

gleb-cloudius replied:
> so I think all this complexity will not be needed.

We still need to be able to transfer a snapshot if there is a slow follower and a snapshot is taken while cdc generation is transferred. IMO it just shows that we should use storage proxy for that.

kbr-scylla replied:
> We still need to be able to transfer a snapshot if there is a slow follower and a snapshot is taken while cdc generation is transferred.

Yup. Besides, relying on nodes not joining group 0 during bootstrap is not quite robust - what if we allow concurrent bootstraps in the future?

> IMO it just shows that we should use storage proxy for that.

I disagree.

gleb-cloudius replied:
> > We still need to be able to transfer a snapshot if there is a slow follower and a snapshot is taken while cdc generation is transferred.
> 
> Yup. Besides, relying on nodes not joining group 0 during bootstrap is not quite robust - what if we allow concurrent bootstraps in the future?

Concurrent bootstrap will join `n` nodes simultaneously, not add other nodes during ongoing operation.

> 
> > IMO it just shows that we should use storage proxy for that.
> 
> I disagree.

Considering that according to the logic I described streaming will not be needed what is your justification for a more complicated solution?

gleb-cloudius replied:
And SP will do it more efficiently as well as it knows not to send a mutation to each node in another DC.

kbr-scylla replied:
>  a more complicated solution?

But it's your solution which is more complicated, not mine. I suggest we put everything in group 0, you suggest to put part of cdc outside group 0 and have another part that lives inside group 0 depends on the outside part, and you claim this is less complicated?

kbr-scylla replied:
> Considering that according to the logic I described streaming will not be needed

How so?

gleb-cloudius replied:
> > Considering that according to the logic I described streaming will not be needed
> 
> How so?

Because no node will join while another node is joining already. It means that the partial data will never have to be streamed.

gleb-cloudius replied:
> > a more complicated solution?
> 
> But it's your solution which is more complicated, not mine. I suggest we put everything in group 0, you suggest to put part of cdc outside group 0 and have another part that lives inside group 0 depends on the outside part, and you claim this is less complicated?

Yes. You need to write the data to all node. I propose to use existing infra (mutation_write_handlers) to do so. Scylla is all about writing data to nodes.

kbr-scylla replied:
>  It means that the partial data will never have to be streamed.

This "partial data" becomes "full data" once we write all of it successfully and we commit the CDC generation ID using `topology_change`. Then it will have to be streamed.

kbr-scylla replied:
> I propose to use existing infra (mutation_write_handlers) to do so. 

you're contradicting yourself from the other thread:

>    I propose to write the code similar to what we have for batches in storage_proxy now. It will construct specially crafted mutation_write_handler with all the targets that we want to write into.

gleb-cloudius replied:
> > It means that the partial data will never have to be streamed.
> 
> This "partial data" becomes "full data" once we write all of it successfully and we commit the CDC generation ID using `topology_change`. Then it will have to be streamed.

After that point it is transferred as part of the topology snapshot. I do not propose to change that.

kbr-scylla replied:
Let's discuss this on todays daily

gleb-cloudius replied:
> > I propose to use existing infra (mutation_write_handlers) to do so.
> 
> you're contradicting yourself from the other thread:
> 
> > I propose to write the code similar to what we have for batches in storage_proxy now. It will construct specially crafted mutation_write_handler with all the targets that we want to write into.

Those two statement are literally identical.

margdoc replied:
Summarized in https://github.com/scylladb/scylladb/pull/13962#issuecomment-1609112962

piodul commented on test/topology_experimental_raft/test_cdc_generation_data.py:
```diff
@@ -0,0 +1,22 @@
+    first_server = await manager.server_add(config=config)
+
+    async with inject_error(manager.api, first_server.ip_addr, 'cdc_generation_mutations_replication'):
+        await manager.server_add(config=config)
```
The test works under the assumption that the error injection properly inflates the size of the CDC generation and grows it beyond the size of the commitlog segment. If that breaks and the inflated size won't be big enough, then the test will stop being useful.

I think you should at least assert that the CDC generation mutation was sent over multiple commands. You could look at the `system.group0_history` table and look at the labels there. The first and the last of the commands that transfer a CDC generation mutation should have a non-empty label already; you could just see whether there is a command of form `"insert CDC generation data (UUID: {}), part 1/{}"` there.

margdoc replied:
Done.

kbr-scylla commented:
> Sounds like a disaster.

@avikivity I think @margdoc forgot to add, that the partially inserted generation data in `CDC_GENERATIONS_V3` won't be used by anything because it won't be pointed at by anything. It's just a partially inserted partition that nobody uses.

This is similar to what happens today: we first insert CDC generation data into `CDC_GENERATIONS_V2` and then we start gossiping the new generation's ID. But if we crash before gossiping reaches other nodes, the generation won't be used, but the data will stay in the table (doing nothing).

kostja commented on service/raft/group0_state_machine.cc:
```diff
@@ -107,7 +124,11 @@ future<> group0_state_machine::apply(std::vector<raft::command_cref> command) {
+           co_await write_mutations_to_database(_sp, cmd.creator_addr, std::move(chng.mutations));
+           co_await _ss.topology_transition(_cdc_gen_svc);
+        },
+        [&] (write_mutations& muts) -> future<> {
+           return write_mutations_to_database(_sp, cmd.creator_addr, std::move(muts.mutations));
```
@margdoc please resolve this as this seems to be fixed.

margdoc replied:
Fixed.

kbr-scylla commented:
@margdoc @kostja @gleb-cloudius I think we managed to reach consensus on the daily.

Instead of keeping `in_progress_cdc_generation_id` field in `topology`, we will send the entire contents of `CDC_GENERATIONS_V3` as part of the group 0 snapshot.

As a follow-up, we'll introduce a mechanism to remove old CDC generations. One piece of the solution will be to make `CDC_GENERATIONS_V3` a single-partition table, turning the generation ID into a timeuuid and making it a clustering key of the table so we can efficiently remove generations with IDs older than X for some carefully picked X.

But that's for a follow-up; for now let's simplify this PR to just send everything.

margdoc commented:
v8:
* rebased on master
* changed sequential sending of write_mutations to parallel sending
* changed snapshotting (https://github.com/scylladb/scylladb/pull/13962#issuecomment-1609112962)
* added check that injection work properly at the of the test
* added description argument to `add_entry_unguarded`
* added release_guard
* fixed commit message 
* fixed compilation of the commit with topology_change refactor

margdoc commented on service/raft/raft_group0_client.cc:
```diff
@@ -469,5 +472,7 @@ void raft_group0_client::set_query_result(utils::UUID query_id, service::broadca
 template group0_command raft_group0_client::prepare_command(schema_change change, group0_guard& guard, std::string_view description);
 template group0_command raft_group0_client::prepare_command(topology_change change, group0_guard& guard, std::string_view description);
+template group0_command raft_group0_client::prepare_command(broadcast_table_query change);
+template group0_command raft_group0_client::prepare_command(write_mutations change);
```
Done.

margdoc commented on service/storage_service.cc:
```diff
@@ -743,6 +751,12 @@ topology_mutation_builder& topology_mutation_builder::set_new_cdc_generation_dat
+topology_mutation_builder& topology_mutation_builder::set_in_progress_cdc_generation_id(
+        const utils::UUID& value) {
+    _m.set_static_cell("in_progress_cdc_generation_id", value, _ts);
+    return *this;
```
This part of the code was removed due to a change of topology snapshots. Closing conversation.

margdoc commented on service/storage_service.cc:
```diff
@@ -757,6 +771,13 @@ topology_mutation_builder& topology_mutation_builder::del_global_topology_reques
+    auto cdef = _s->get_column_definition("in_progress_cdc_generation_id");
+    assert(cdef);
+    _m.partition().static_row().apply(*cdef, atomic_cell::make_dead(_ts, gc_clock::now()));
+    return *this;
```
This part of the code was removed due to a change of topology snapshots. Closing conversation.

margdoc commented on idl/group0_state_machine.idl.hh:
```diff
@@ -27,8 +27,12 @@ struct topology_change {
     std::vector<canonical_mutation> mutations;
 };
 
+struct write_mutations {
```
Fixed

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2245/

gleb-cloudius commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2245/

You need to amend command size in cmd merging test since you changed the max mutation size.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -1056,6 +1049,49 @@ class topology_coordinator {
         co_return std::pair{gen_uuid, std::move(gen_mutations)};
     }
 
+    // Broadcasts all mutations returned from `prepare_new_cdc_generation_data` except the last one.
```
The commit message and this comment need to be updated

margdoc replied:
Fixed.

kbr-scylla commented on test/topology_experimental_raft/test_cdc_generation_data.py:
```diff
@@ -0,0 +1,31 @@
+        if row.description.startswith('insert CDC generation data (UUID: ') and row.description.endswith('), part'):
+            break
+    else:
+        assert False, "No CDC generation data sent in parts was found"
```
The standard way to do this is `pytest.fail(...)`

margdoc replied:
TIL, thanks.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5408,11 +5410,12 @@ void storage_service::init_messaging_service(sharded<service::storage_proxy>& pr
 
-            std::optional<canonical_mutation> cdc_generation_mutation;
+            std::vector<canonical_mutation> cdc_generation_mutations;
             if (curr_cdc_gen_id) {
```
I don't think you want to do this check now.
There may be no current CDC generation, but one generation is in the middle of being sent.
If you triggered a snapshot transfer at that moment, this condition would be `false` and the receiver wouldn't obtain the partially applied generation.

It looks like we can get rid of `curr_cdc_gen_id` variable altogether.

margdoc replied:
Fixed.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5408,11 +5410,12 @@ void storage_service::init_messaging_service(sharded<service::storage_proxy>& pr
-            std::optional<canonical_mutation> cdc_generation_mutation;
+            std::vector<canonical_mutation> cdc_generation_mutations;
             if (curr_cdc_gen_id) {
                 // We don't need to fetch this data under group0 apply mutex, it's immutable.
```
"immutable" was true when referring to the current CDC generation which we would only transfer once it was fully applied.

Now we transfer the entire contents of CDC_GENERATIONS_V3, which may contain a partially applied generation data which we are concurrently applying more of. So it is no longer immutable.

However, it may be that it is still safe to pull such generations that are being concurrently applied -- but now the reason is different. If you found such a reason and convinced yourself that it is safe, then you should also convince us (by updating the comment appropriately).

margdoc replied:
I'm not 100% sure it is safe, so I will add this mutex here.

margdoc replied:
Fixed.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5408,11 +5410,12 @@ void storage_service::init_messaging_service(sharded<service::storage_proxy>& pr
-                // because nodes only load whatever `current_cdc_generation_id` points to in topology.
+                // We only need to include the current CDC generation data in the snapshot
+                // and the generation of data is being broadcasted but hasn't been entirely
+                // applied yet.
```
I would get rid of this last sentence, it was useful to explain this when we were only pulling the current generation, now we're pulling everything so it's unnecessary.

margdoc replied:
Fixed.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5423,22 +5426,26 @@ void storage_service::init_messaging_service(sharded<service::storage_proxy>& pr
-                auto key = dht::decorate_key(*s, partition_key::from_singular(*s, curr_cdc_gen_id->id));
-                auto partition_range = dht::partition_range::make_singular(key);
+                //
+                // FIXME: Topology snapshots contain only mutation of current CDC generation data but don't
```
"Topology snapshots contain only mutation of current CDC generation data"
But that's not true (after your changes)

margdoc replied:
Fixed.

kbr-scylla commented on service/raft/group0_state_machine.cc:
```diff
@@ -91,7 +91,8 @@ future<> group0_state_machine::apply(std::vector<raft::command_cref> command) {
         merger(group0_state_machine& sm_, utils::UUID id, semaphore_units<> mux) : last_group0_state_id(id)
```
commit message:
"i`max_mutation_size`" (what's the "i" for)
"Fixe" -> "Fixed"

I would start the sentence "A batch of raft commands has the size at most..." on a new paragraph, so it's clearer that it talks about a different thing (the code inside `group0_state_machine`) than the previous part (which was talking about the code in `raft_sys_table_storage`).

margdoc replied:
Fixed.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -1595,6 +1595,8 @@ future<> topology_coordinator::run() {
             }
         } catch (raft::request_aborted&) {
             slogger.debug("raft topology: topology change coordinator fiber aborted");
+        } catch (raft::commit_status_unknown&) {
```
Please explain in the commit message in what situations we currently expect `commit_status_unknown`. Before your PR it should not have happened, but now you introduced the possibility of it happening so we expect it and it's not an error.

margdoc replied:
Done.

margdoc commented:
v9:
* fix comments after changing sequential to parallel sending `write_mutations`
* change the constant number in `boost.group0_cmd_merge_test`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2269/

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request

margdoc commented:
v10:
* fixed comments and commits messages
* always send entire CDC generation data in topology snapshot

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5425,22 +5420,22 @@ void storage_service::init_messaging_service(sharded<service::storage_proxy>& pr
-                auto key = dht::decorate_key(*s, partition_key::from_singular(*s, curr_cdc_gen_id->id));
-                auto partition_range = dht::partition_range::make_singular(key);
+                //
+                // For now, this problem is temporarily solved by sending the entire CDC_GENERATIONS_V3 table.
```
Not exactly -- we are still only fetching `current_cdc_generation` from the table into memory on nodes that apply topology commands or group 0 snapshot.

You did improve the situation with your PR -- the nodes are now **receiving** all generations, however, they still only looking at the `current` one.

So you could improve the comment above: instead of "The last bootstrapped node will only receive the newest CDC generation...", write something like "The last bootstrapped node will only read the newest CDC generation into memory...".

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request

margdoc commented:
v11:
* fixed typo in commit message
* fixed comment about bootstrapped nodes

kbr-scylla commented:
CI: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2276/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2276/

kbr-scylla commented:
> CI state FAILURE - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2276/

Looks like the second node booted by the test hanged due to topology coordinator errors:
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2276/artifact/testlog/x86_64/debug/scylla-147.log
```
WARN  2023-07-05 16:16:30,396 [shard 0] storage_service - raft topology: got stream_ranges request while my tokens state is {} and node state is none
WARN  2023-07-05 16:16:30,488 [shard 0] storage_service - raft topology: got stream_ranges request while my tokens state is {} and node state is none
...
```

kbr-scylla reviewed (changes requested):
Please fix the order of commits due to the reasons explained by Piotr here: https://github.com/scylladb/scylladb/pull/13962#discussion_r1241261644

and please investigate the failure.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1056,6 +1049,51 @@ class topology_coordinator {
+        if (updates.size() > 1) {
+            release_guard(std::move(guard));
+
+            co_await parallel_for_each(updates.begin(), std::prev(updates.end()), [this, gen_uuid = gen_uuid] (canonical_mutation& m) {
```
Why can't we send all mutations including the last one? If term changes we will fail to write the topology mutation. We do not need cdc mutation for that.

margdoc replied:
If we send the last mutation in the `write_mutation` command, we would use a total of `n + 1` commands instead of `n-1 + 1` (where n is the number of mutations), so it's better to send it in `topology_change` (we need to send it after all `write_mutations`) with some small metadata (`transition_state`, `new_cdc_generation_data_uuid`, etc.). There is no reason for excluding the last mutation here, other than the optimization of a number of raft commands sent.

gleb-cloudius replied:
Not sure the optimization worth the complication, but lets at least amend the commit message and may be write a comment above the function that we want the happy path to issue only one raft command instead of two.

margdoc commented:
v12:
* added paragraph to `prepare_and_broadcast_cdc_generation_data` commit's message
* fixed `merger` bug
* reordered commits

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2333/
